Fixes Authorization for 3.x

bravo-kernel opened the pull request:
Please review, if this is OK I will attempt to update the tests.

- supports the INI convention as [described here](https://gist.github.com/bravo-kernel/2a4c5df8785671ce86de)
- tested successfully for all scenarios in the INI above (plugins, prefixes, etc)
- major difference; I chose to generate a slightly more describing array for the acl since that could be useful for debugging, reporting, etc

Feedback welcome

dereuromark commented on src/Auth/TinyAuthorize.php:
```diff
@@ -60,7 +60,8 @@ class TinyAuthorize extends BaseAuthorize {
 	 */
-	public function __construct(ComponentRegistry $registry, array $config = array()) {
+	public function __construct(ComponentRegistry $registry, array $config = array())
+	{
```
Please keep the PSR-2-R cs.

bravo-kernel replied:
np, I thought Cake was following PSR-2 with that damned newline curly  bracket.

Do I understand correctly that you want it on the same line?

dereuromark replied:
See the badge in the readme: [![Coding Standards](https://img.shields.io/badge/cs-PSR--2--R-yellow.svg)](https://github.com/php-fig-rectified/fig-rectified-standards)

dereuromark commented on src/Auth/TinyAuthorize.php:
```diff
@@ -157,21 +167,20 @@ public function validate($roles, $plugin, $controller, $action) {
-			if (array_key_exists($controller, $this->_acl) && !empty($this->_acl[$controller][$action])) {
-				$matchArray = $this->_acl[$controller][$action];
+		// allow access if user has been granted access to the specific resource
+		$action = Inflector::underscore($request->action);
```
Why do we need to inflect?
Some people use the new "myActionName()" method syntax. So the ini file should then also have that exact naming scheme IMO for the actions.

bravo-kernel replied:
ok, will remove the inflection

bravo-kernel commented:
The tests are now passing but I only updated the existing ones. I will add some more this weekend. Is the overall approach ok?

dereuromark commented on src/Auth/TinyAuthorize.php:
```diff
@@ -275,4 +283,67 @@ protected function _getAcl($path = null) {
+	 * @param string INI section key as found in acl.ini
+	 * @return string String converted to use cake conventions
+	 */
+	protected function normalizeIniKey($key) {
```
protected functions should have the _ underscore for consistency.

bravo-kernel replied:
Changed

dereuromark commented on src/Auth/TinyAuthorize.php:
```diff
@@ -275,4 +283,67 @@ protected function _getAcl($path = null) {
+	 * @return array Hash with named keys for controller, plugin and prefix
+	 */
+	protected function constructIniKey(Request $request) {
+		$res = Inflector::camelize($request->params['controller']);
```
I wonder if we need the inflections here.

bravo-kernel replied:
I thinks there's a choice to be made here; 

1.force users to be absolutely correct when defining their acl.ini entries (casing)
2. use Inflections (as in the code above) so e.g. accounts.accounts in the acl.ini gets converted to Accounts.Accounts. 

Same goes for the acl.ini you commented on, I used lowercase 'users' intentionally to test the conversion.

Anyway, either way is fine by me. Just let me know what you prefer so I can round this up.

dereuromark replied:
I think they should be 100% correctly cased.

If we would blacklist, then it would be a security risk if you misspelled and didnt inflect. But since we whitelist, the worst case scenario would be missing access to an action.

And the 3.x core in general also goes to less inflection and more accurancy (for speed and clarity).

bravo-kernel replied:
yes sir :+1:

dereuromark commented on tests/TestCase/Auth/TinyAuthorizeTest.php:
```diff
@@ -39,8 +39,14 @@ public function setUp() {
-admin_index = admin
+;edit = user
+;* = admin
+[users]
```
Should be "Users" - camlized U

dereuromark commented on tests/TestCase/Auth/TinyAuthorizeTest.php:
```diff
@@ -81,27 +92,51 @@ public function testConstructor() {
+				'controller' => 'Tags',
+				'actions' => [
+					'add' => [1, 2, 3, -1],
+					'very_long_action_name_action' => [1],
```
veryLongActionNameAction?

bravo-kernel replied:
I'm not sure to be honest, it was in the existing tests so I just left the test in there.

dereuromark replied:
I think it depends on if a user uses Dashed or Inflected route. for the latter it is fine, for Dashed, the actions must be camelBacked AFAIK. Correct me if I am wrong.

bravo-kernel commented:
One more thing came up from the tests:

The ``allowUser`` description now describes that enabling it gives users access to all non-admin-prefixed routes. It needs some updating to get that working right and I suggest that enabling it gives users access to all actions not NOT using the configurable ``adminPrefix``. Gives the same result, access to everything but a bit more flexible.

What  do you think?

dereuromark commented:
I agree with that.

bravo-kernel commented:
OK, one more thing besides the dashed/inflected route:

The tests in https://github.com/bravo-kernel/cakephp-tinyauth/blob/cake3-prefixing/tests/TestCase/Auth/TinyAuthorizeTest.php#L670 no longer make sense to me. They seems to do the exact same tests as used for normal access.

Am I missing something?

dereuromark commented:
Yeah lets remove that.

bravo-kernel commented:
OK, I think this should be good now as far as the core logic is concerned, test cases are added for all the scenarios I could think and things look quite solid if you ask me.

If will need some help on how to set up the tests for testing dashed vs. inflected routes.

bravo-kernel commented:
I have added a base document ``authorize.md`` accompanying the change. i have not linked to it from the README since I believe you probably have your own ideas about how the structure will have to look like.

As far as I'm concerned this should be good enough to merge now. Let me know!

ps: I changed the location of the constants to be in ``app_custom.php`` since the 2.x suggestion to to define them in ``bootstrap.php`` does not seem to work (anymore?). The config is loaded before bootstrap nowadays.

dereuromark commented on docs/authorize.md:
```diff
@@ -0,0 +1,127 @@
+	'Roles' => [
+		'user' => ROLE_USERS,
+		'admin' => ROLE_ADMINS,
+		'supergirls' => ROLE_SUPERADMINS
```
superadmin? :) and SUPERADMIN (singular)

bravo-kernel commented:
One more thing, there are two checks against a ``-1`` role id are performed and I don't think they are needed anymore.

- https://github.com/dereuromark/cakephp-tinyauth/blob/master/src/Auth/TinyAuthorize.php#L150
- https://github.com/dereuromark/cakephp-tinyauth/blob/master/src/Auth/TinyAuthorize.php#L166

dereuromark commented:
I agree :) We removed that feature so that can go now.

bravo-kernel commented:
Not quite happy with the tests yet, they need to be cleaned up and a bit more verbose for (future) maintenance. One more update will follow.

bravo-kernel commented:
I have cleaned up the tests, added verbosity to (hopefully) make future maintenance easier, made the tests look consistent and have consolidated the acl.ini used for the testcases. 

I honestly would not know what more could be added at this point ;)

bravo-kernel commented:
Could you merge? I would like to continue with HABTM
